Implement the RoutesDataClass as a generic data format

chenyangkang opened the pull request:
- [x] Add `Routes`, `BirdFlowRoutes`, `BirdFlowIntervals` #202 
- [x] Add interval validation for data classes.
- [x] Add conversion functions and helper functions for conversion among the classes. (resolves #158)
- [ ] Add generic S3 methods for the three classes:
  - [x] `print`
  - [ ] `plot ` or `plot_routes`?
- [x] Add initial tests for the three data classes
- [ ] Add complete tests for the three data classes
- [ ] Update all components in BirdFlowR regarding the generic data format. Use the new data classes to replace them:
  - [ ] `route` in `route.R`
  - [ ] `format_trajectory` in `route.R`
  - [ ] `snap_to_birdflow` in `snap_to_birdflow.R` (replace it or keep it)?
  - [ ] `calc_predictive_distance_metric.R`
  - [ ] `interval_log_likelihood.R`
  - [ ] `make_tracks` in `function.R` of `BirdFlowPipeline` package.
  - [ ] `preprocess_with_recovery`, `preprocess_sort_band_date`, `preprocess_calc_distance_days` in `function.R` of `BirdFlowPipeline` package.
  - [ ] `tracking_data.R` in `BirdFlowPipeline` package.
  - [ ] `amwo_track_info_new.R` in `BirdFlowPipeline`
  - More to add
  

Minor changes:
- [x] Remove bridflow_crs and related. #203 
- [x] Add functions (initial) for calculating distance metrics #201

chenyangkang commented:
Don't know why `test-preprocess_species.R` throw an error. All test passed locally.  `expect_true(all((ext(a)[, ] %% xres(a)) < 1e-9))` This line throws an error, likely a resolution problem.

chenyangkang commented:
Local:

```
> a <- preprocess_species("example_data", hdf5 = FALSE)
The example dataset does not represent a complete species range
so should only used for demonstrating package functions.
Species resolved to: 'yebsap' (Yellow-bellied Sapsucker)
Downloading Status Data Products for yebsap-example
Data already exists, use force = TRUE to re-download.
Setting max_params to 278697612 anticipating 12 GB of GPU ram.
Calculating resolution
  Attempt 1 at setting resolution
  (6.799km chosen)
  244.76% of target (estimate).
  trying again
  Attempt 2 at setting resolution
  (8.812km chosen)
  85.98% of target (estimate).
  trying again
  Attempt 3 at setting resolution
  (8.522km chosen)
  99.11% of target (estimate).
 success
Rounded to9km final resolution.
Resolution forced to 30 for example data,which only has low resolution images
process_rasters called with download_species =yebsap-example
Reading low resolution (27 km) geoTIFFs
Creating mask in target resolution and projection
Reprojecting and cropping to mask:
	abundance done.
	Upper CI done.
	Lower CI done.
Model has:
	372 active cells,
	52 transitions, and
	2,795,732 parameters
0.1gb of GPU ram required to fit model.
> a
Yellow-bellied Sapsucker BirdFlow model
  dimensions   : 24, 27, 52  (nrow, ncol, ntimesteps)
  resolution   : 30000, 30000  (x, y)
  active cells : 372
  size         : 1.1 Mb
> a
Yellow-bellied Sapsucker BirdFlow model
  dimensions   : 24, 27, 52  (nrow, ncol, ntimesteps)
  resolution   : 30000, 30000  (x, y)
  active cells : 372
  size         : 1.1 Mb
> xres(a)
[1] 30000
> ext(a)
SpatExtent : 690000, 1500000, 690000, 1410000 (xmin, xmax, ymin, ymax)
> ext(a)[, ]
   xmin    xmax    ymin    ymax 
 690000 1500000  690000 1410000 
> ext(a)[, ] %% xres(a)
        xmin         xmax         ymin         ymax 
0.000000e+00 0.000000e+00 1.164153e-10 2.328306e-10 
```

chenyangkang commented:
@ethanplunkett Hi Ethan, could you please help review this bug at your convenience?

chenyangkang commented:
Seems to be a floating-point precision errors induced by `%%` operation. I changed it in another way using `%/%` and the problem seems to be solved. Very strange, might depend on system version.

chenyangkang commented:
@ethanplunkett Hi Ethan do you want to review the current changes before I go further to edit other functions? Especially about `Routes`, `BirdFlowRoutes`, `BirdFlowIntervals` (in RoutesDataClass.R), `as_BirdFlowRoutes`, `as_BirdFlowIntervals` (in RoutesDataFunction.R).

I still implemented the transformation function `as_*`, and kept the class functions open to dataframes that are already in the right format (like the dataframe generated by `format_trajectory` in the `route` function).

ethanplunkett commented:
@chenyangkangyes.  I want to review before it's merged.  I have come down with something nasty though so am unlikely to get to it for a few days.  I'm currently running a 102 fever.  

I do have a couple thoughts based on what the code looked like Saturday morning when O last looked it over.  Disregard anything that you've already addressed.  


1. Ideally we'd have no new Notes in the package check.  There were a couple that predate this update that you are unlikely to address. 

2. When you align to birdflow while converting to 'BirdFlowRoutes' use 'snap_to-birdflow()' as that checks validity against the dynamic mask using the date and time simultaneously. It looked like your code was checking them independently which would allow locations that are excluded by the dynamic mask as long as they are in the static mask.  I would expose the 'aggregate' argument too and add a 'random' option to mimic the default behavior of the new function. 

3.  We need to make sure 'plot_routes()' works with the modified routes.  It's a bit of a mess so and the mess is all mine so let me know if you want me to clean it up.

chenyangkang commented:
@ethanplunkett Sorry to hear that! And thanks for the quick feedback. I will check the first two points.

Regarding the `plot_routes` function, it seems to be working fine with the new class -- The data structure is not changed. Not sure if some hidden issue exists about that "crossing boundary" thing. I attached it and let me know if it doesn't look right. Also I'm thinking how about transforming (renaming) the `plot_routes` to a generic `plot` function on the `BirdFlowRoutes` and `Routes` class.

Hope you get better soon and we can discuss later!

<img width="586" alt="image" src="https://github.com/user-attachments/assets/5e90f36a-7476-41ac-9a44-614553c10221" />

ethanplunkett commented:
Yes. It’s a good idea to  make plot routes a method of the generic ‘plot’ now that we’ve formalized the class structure.  On Jan 20, 2025, at 5:13 PM, Yangkang ***@***.***> wrote:﻿
@ethanplunkett Sorry to hear that! And thanks for the quick feedback. I will check the first two points.
Regarding the plot_routes function, it seems to be working fine with the new class -- The data structure is not changed. Not sure if some hidden issues exist about that "crossing boundary" thing. I attached it and let me know if it doesn't look right. Also I'm thinking how about transforming (renaming) the plot_routes to a generic plot function on the BirdFlowRoutes and Routes class.
Hope you get better soon and we can discuss later!
image.png (view on web)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

ethanplunkett commented on R/index_conversions.R:
```diff
@@ -60,7 +60,7 @@
 #'  space index corresponding to x and y coordinates or row and column indices.}
 #'  }
 #'
-#'  \item{`latlon_to_xy(lat, lon, bf)`}{Returns a two column matrix of the x
```
This was right as it was `bf` is an argument to `latlon_to_xy()`

ethanplunkett commented on R/is_between.R:
```diff
@@ -33,7 +33,7 @@ if (FALSE) {
 #' @param radius A point is considered between two locations if it is within
 #' `radius` meters (along a great circle) of the great circle line between the
 #' locations. `radius` defaults to half the cell size (`mean(res(bf))/2`).
-#' @param n_direction The number of (equally spaced) directional bins to
```
Thanks for fixing this!

ethanplunkett commented on R/route.R:
```diff
@@ -173,15 +173,26 @@ format_trajectory <- function(trajectory, bf, timesteps) {
+  points <- points |>
+      dplyr::group_by(route_id) |>
+      dplyr::mutate(
+          date = as.Date(unlist(purrr::accumulate(date, ~ ifelse(.y < .x, .y + lubridate::years(1), .y))))
```
Currently synthetic routes are allowed to be backwards in time.  This assumes forward.  We should either change `route()` to exclude backwards routes or clean things up here.  `purr` is iterating through each pair or adjacent rows so is likely to be slow.  It might make sense to fix this in `format_trajectory()`.

ethanplunkett commented on R/validate_RouteDataClass.R:
```diff
@@ -0,0 +1,392 @@
+  # check the features required by direct initiation of BirdFlowIntervals class
+  for (name in get_target_columns_BirdFlowIntervals(type='input')){
+    if (!name %in% colnames(birdflow_interval_df)){
+      stop(sprintf(glue::glue("'{name}' is not found in the input dataframe.")))
```
Drop dependency on glue and use `paste()` or `sprintf()`

ethanplunkett commented on R/validate_RouteDataClass.R:
```diff
@@ -0,0 +1,392 @@
+
+#' @rdname target_columns
+#' @export
+get_target_columns_Routes <- function(type='input'){
```
These names feel clunky.  I'd be tempted not to export them as I don't think they are for end users, but if we do want to export them I think they should be shorter.  `get_routes_cols()`  `get_birdflowroutes_cols()` ?   There's probably something better.

ethanplunkett reviewed (commented):
@chenyangkang   I left lots of specific comments suggesting changes but overall this is great!   

Some overarching comments:
*  Stylistically I prefer more succinct names. I don't think that matters enough to change it for the private functions and internal code, but it is worth shortening the public facing function and argument names. 
* I don't remember deciding whether we wanted to base these classes on data frames or lists.  I do remember saying data frames might be cool.  I like that they can then be treated like data frames but as you discovered manipulating them can lose the attributes, a problem people are less likely to bump up against with a list based object.  Let's talk about this in person; I don't have a strong feeling either way but want to make sure we are making an informed decision.

ethanplunkett commented on tests/testthat/test-preprocess_species.R:
```diff
@@ -13,7 +13,22 @@ test_that("preprocess_species runs on test dataset", {
   expect_no_error(a <- preprocess_species("example_data", hdf5 = FALSE))
   expect_no_error(validate_BirdFlow(a, allow_incomplete = TRUE))
   expect_error(validate_BirdFlow(a))
-  expect_true(all((ext(a)[, ] %% xres(a)) < 1e-9))  # Test if origin is at 0, 0
```
This test is making sure the cells align with the origin and I think is important to keep.   It has nothing to do with your code.

I have two thoughts:
1.  I could add code to preprocess species to check if the extent and cell size numbers are close to integers and round them if they are. It wouldn't help when the cell size is floating point (say some sort of spherical coordinates) but would help with all practical applications we've had to date.
2.  We could lower the tolerance on the check to allow the test to pass as is.  I can't see the failed test output anymore but if I recall the difference was very small but not quite as small as this test requires.

chenyangkang replied:
I think I will leave the current modification as it is -- To suppress the error. I'm not deleting the test -- just using another logic to validate.

ethanplunkett replied:
Sorry I missed the added lines!

Yours is definitely better.

Your test passes if you are within the tolerance either above or below the exact number, whereas mine only passes if you are above but within the tolerance.  
 What's weird is when it was failing before it was for a value that was slightly too big given the tolerance so, I think, should have failed either test.   Definitely leave yours though, it's both theoretically and empirically working better. 

Thanks!

ethanplunkett commented:
> @ethanplunkett Hi Ethan, could you please help review this bug at your convenience?  (Failed test for cell alignment unrelated to this PR).
I made a comment here: 
    https://github.com/birdflow-science/BirdFlowR/pull/204/files#r1925521860
I'll try to fix it in another branch.   It's unrelated to your code.

ethanplunkett commented:
RE: `plot_routes()`  that 31 week stay looks suspicious to me.

chenyangkang commented:
> RE: `plot_routes()` that 31 week stay looks suspicious to me.

Oh I see, the elapsed_stay is calculated by row_number int he plot_routes function -- Will change it. It should represent 'days' now.

chenyangkang commented on R/RouteDataFunction.R:
```diff
@@ -0,0 +1,637 @@
+add_stay_id_with_varied_intervals <- function(df, timestep_col = "date", timediff_unit = "days", time_threshold = Inf) {
+  
+  # Ensure the data is sorted by timestep
+  df <- df |> dplyr::arrange(.data[[timestep_col]])
```
Indeed it can cause problem when applying to a dataframe containing multiple routes. But when using it in the `new_BirdFlowRoutes` will first groupby route_id, and then do this operation for each group -- so avoid the problem. I will add annotation to this function that it should only be applied on a single route.

```R
new_BirdFlowRoutes <- function(data, species, metadata, geom, dates, source, 
                               stay_calculate_col = "date", stay_calculate_timediff_unit = "days") {

  ## Add stay id
  data <- data |>
    sort_by_id_and_dates() |>
    dplyr::group_by(.data$route_id) |>
    add_stay_id_with_varied_intervals(timestep_col = stay_calculate_col, timediff_unit = stay_calculate_timediff_unit) |> 
    # Here, using add_stay_id_with_varied_intervals, rather than add_stay_id. 
    # It takes 'timestep' as input so account for varying intervals, 
    # if the data is not sampled in a frequency.
    dplyr::ungroup() |>
    as.data.frame()
   
 ...
```

chenyangkang commented on R/RouteDataFunction.R:
```diff
@@ -0,0 +1,637 @@
+    dplyr::mutate(
+      timestep_diff = c(1, as.numeric(diff(.data[[timestep_col]]), units = timediff_unit)),  # Time differences
+      i_change = c(1, as.numeric(diff(.data$i)) != 0),    # Changes in 'i'
+      stay_id = cumsum(i_change | (timestep_diff > time_threshold))
```
Yes. My thoughts is that if this function is used for plotting real tracking data, there should be a function that bin the data into spatiotemporal grids and remove redundancy -- Just like what `to_BirdFlowRoutes` does. In that case this function will function correctly.

ethanplunkett commented:
Looks like you are close!  

This is great. 

I have one more comment. 

Can you add to the top of the news the way this breaks existing code. Something along the lines of:

Breaking Changes:

1. Old route objects will no longer plot with the updated plot_routes()

2. Anyone who is using plot_routes() to plot data not generated with route() will have to update their objects.  [describe how]

chenyangkang commented:
@ethanplunkett Done! Could you please help review that next week at your convenience? Ready to be merged.

Still need to do several things:
1. Go through the historical functions and update the code using the new classes.
2. Add a plot function for class `Routes` so that routes in non-BirdFlow coordinates can also be plotted. I think we can use a default CRS and spatiotemporal cell size as argument for that function.

-- These will go to other PRs. I think the current changes are all for this PR.

ethanplunkett commented:
# Changes I made to the PR

## Changed validation / object definitions

* geom component must be fully defined and valid -
* species component must have as a minimum the `common_name` element, can only have elements in 
      `new_birdflow()$species`  also when creating a routes object you can provide a single element species which is passed to `ebirdst::get_species()` to set full species information.
* `source`  must be `character` or `NA`.  It can be a vector.
* `BirdFlowRoutes` must have two specific metadata items (needed for plotting to work).
* `snap_to_birdflow()` now uses the time component of date time objects. Previously everything was converted to dates.  
* `snap_to_birdflow()` Now returns date-times if input is date-times.  
* `Route` and `BirdFlowRoute` functions now work with backwards routes.
* `as_BirdFlowRoutes()` now pulls metadata from the BirdFlow object.
* Move `stay_units` argument to last position in `plot_routes()` to avoid breaking any existing code that relies on argument position. 
 
## Reorganizing 
* Almost all functions are now private with the exception of `Routes()` `as_BirdFlowRoutes()` and `as_BirdFlowIntervals()`  
* Definitions and tests for these three functions moved to independent files.
* Most other files now in `Routes-internal`.

## Updated generic functions

* `plot_routes()` (and thus `plot()`) now works with `BirdFlowRoutes` and `Routes`
* `animate_routes()` works.
* dropped plot method for `BirdFlowRoutes`
* plot method for `Routes` now used for both `Routes` and `BirdFlowRoutes` due to inheritance.
* Use dates not timesteps in `plot_routes()`
* Updated plotting for all three of the new classes.
    * Only print first 5 lines of `data`
    * Print `data` last.
    * Don't print non-essential stuff if it's not defined
    * Only print counts by type if there's more than one type
    * Don't print `metadata`

## Fixed bugs

* functions now honor the `sort_by_id_and _dates` argument.  Previously if set to FALSE sorting 
still took place.
* Fixed plotting issues
* As `snap_to_birdflow()` and `as_BirdFlowRoutes()` now honor time component of input objects. 

## Updated Testing

* Added `make_fake_tracks()` helper function to generate fake track data with high spatial and temporal resolution.
* Dropped some snapshot tests
* Made some tests more explicit 


# To Do:

## Testing

Drop most if not all snapshot tests and replace with more explicit tests.  Snapshot tests are hard to maintain (e.g. changing the print methods breaks them), and they aren't actually confirming that the function is performing as expected. Instead write the test to present the function with data that looks similar to what you plan to use it on and then write code to explicitly evaluate whether the result is as expected.  I began this process and discovered a couple bugs and some unexpected / poorly defined behavior.

## Bugs

These haven't been fixed.

* `as_BirdFlowIntervals()`   doesn't honor the minimum distances and times input in the 
arguments. This is true even with the test results that are in the original snapshots.  I added to the existing test an expectation that calculated the distance and lag in days for each interval was greater than the minimum in the arguments and this bug jumped out.  I have not fixed this one.

* Stay length - We now calculate the elapsed time based on dates AFTER snapping to birdflow and aggregating all the observations within a week into one, which may be a randomly selected observation or a central observation, or an average date-time.
So, for example, if a bird is in a cell for 6 days that are all within one timestep we will collapse that down to a single date and  call the elapsed time zero, but if it's there for 6 days that span 2 timesteps the result will likely depend on the aggregation method and/or random seed and could range from 6 to 0 days.  We could add first_date and last_date columns to snap_to_birdflow() output and use them to calculate the stay length, or we could snap all dates to the center of the week and live with weeks as our precision. There's a related problem in that `plot_routes()` calculates an `ellapsed_stay` which is similar to stay but changes for each row of the stay. Thus with weekly data in weeks it would increment 1, 2, 3, 4 as the stay progresses. This is used when animating routes.  The problem being that we can't calculate this with precision within the plot function even if we have precise stay values stored in the object.  So, if we are going to have precise stay lengths we should calculate `ellapsed_stay` at the same time and include it as a formal component of a BirdFlowRoutes objects.
See also unresolved time aggregation question below which relates to this.


## Linting and Spell checking

Use `devtools::lint()` to lint the code and `devtools::check_spelling()` to check spelling.

## Validation
 
I think we should ignore this for now as it's not public facing but eventually I would like to consolidate the validation code. 

The nested validation functions obfuscate what's going on. For example `validate_BirdFlowRoutes_metadata()` didn't do anything, but it's existence and calling it from the other validators makes it look like the metadata is being validated.

The validators for Routes BirdFlowRoutes and BirdFlowIntervals should contain most of the validation code. As is dozens of lines of code and documentation for validators that often have only one or two lines and that are removed from their context makes it harder to see what's going on.

Helper functions should be used in validation:
* If the primary function is really long to break up code into logical components or 
* to eliminate duplicated code.  

I added a new `validate_geom()` function that meets both of the criteria.  The `validate_BirdFlowRoutes_geom()` function originally in the PR verified that the list item names were present but didn't check to see if they had complete and consistent values, and most of the examples included in the PR didn't have valid geom and thus created objects that would not plot properly. `validate_geom()` incorporates code that was previously in `validate_BirdFlow()` and is now called to validate the geom component anytime it exists (from `validate_BirdFlow`, `validate_BirdFlowRoutes`, and `validateBirdFlowIntervals`). 


# Open questions

* Do we want `BirdFlowRoutes`  `x`, `y`, `date` values to be snapped to cell and week centers.  Currently they are not and represent input or aggregated values but `i` and `timestep` are inherently snapped to the centers.  I suspect we want to snap to cell centers and weeks but can see arguments for retaining both?

* Should we allow extra columns in `data` ?  Currently extra columns are retained in all see `info` column in some tests.
  1. For Routes objects I think it makes sense to allow extra columns as this allows tracking additional id and field data in the object.
  2. For BirdFlowRoutes it's less clear to me that we should allow extra columns but I'm OK with it. If we do allow it we have to have a way to aggregate it when multiple rows are being combined. This could be as simple as taking the first element BUT in that case we should be careful to document this potentially confusing behavior.
  3. For BirdFlowIntervals I think we should NOT allow extra columns as we are always pulling from two rows and it's entirely possible that they will have different values for those columns.
  Currently extra columns are retained in all three, I'm not sure if for the intervals the value is derived from the first or second point. 

* The camel case class names has been bothering me. It dates back to an original sin of this package in that the "BirdFlowR" name was decided before I got involved.  When I named the `BirdFlow` class I decided to follow the case in the package's name.  However, I kept the camel case out of the user facing functions with the exception of `validate_BirdFlow()` which is exported but not often called by users.  I'm not sure what if anything we should do about it.

ethanplunkett commented:
The updated pull request is failing for two reasons:
1. I didn't update the snapshots as I think we should drop most of them.
2. I added tests that revealed a bug that isn't fixed yet.  
See prior comment for more details.

chenyangkang commented:
General updates:

- Updated tests for as_BirdFlowIntervals. Removed redundant components like snapshots.
- I kept both `interval_log_likelihood.R` and `interval_based_validation.R` (this is a new file). The main difference is that the later function provides more interval based metrics like energy score, distance metric, etc. Also, the latter function use `BirdFlowIntervals` object as input. But since 
- ```
  observations <- BirdFlowModels::[rewbla_observations](https://rdrr.io/pkg/BirdFlowModels/man/rewbla_banding_data.html)
  intervals <- BirdFlowModels::[rewbla_intervals](https://rdrr.io/pkg/BirdFlowModels/man/rewbla_banding_data.html)
  intervals  <- interval_log_likelihood(intervals, observations, bf)
  ```

  Is still referring to this function, I’m keeping this in this PR. We may want to remove `interval_log_likelihood.R` and simply replace it with the latter one in the next PR.

- Spell check
- Fixed all examples.
- Add a 600 second timeout tolerance to the `Polygons.Rmd` vignette. The download of US polygon from the U.S. Census seems to be slow for some reason.
- There is a `res = NULL` default argument in `plot_rotues()` that is not used anywhere in the function, so I deleted it.
- Add a parameter documentation for `@param routes` in `animate_routes`. Previously undocumented.
- Updated NEWS.md
- Fixed most lint issues. Some are because of long variable name. I did not fix them since most of them will be for internal use.


Bugs fixed:
- `as_BirdFlowIntervals()` and `as_BirdFlowRoutes()` bugs fixed. Filter out invalid locations.
- About the stay length. I think this is a problem with real track if I understand it correctly, not with synthetic routes. And I’m not sure I fully understand the problem but I feel like any errors within +- 7 days are acceptable since we use a weekly resolution for BirdFlow. If the user choose to snap the raw data into BirdFlow coordinates then they probably should not expect “precise” stay length. So my thought is that to plot a raw track, whether (1) plot precise lon and lat, without stay calculation (plot(Routes)); or (2) custom inputs of spatial and temporal resolution as arguments, and the raw data will be aggregated based on this.


Open questions
- I would consider lon, lat, and date in the BirdFlowRoutes to reflect the unchanged input information (whether synthetic or raw data), and x, y as direct CRS projection of lon and lat. Some functions are built on the x and y (direct projection) so I would hope to keep them unsnapped.
- I have no problem with restricting the extra columns.

chenyangkang commented:
The failure of CI stems from a change of `ebirdst` function in its development version.

chenyangkang commented:
Congrats!!! A significant milestone 🫡 @ethanplunkett
